python/dgl/dataloading/dataloader.py

```diff
@@ -0,0 +1,638 @@
+    exclude_mode : str, optional
+        Can be either of the following,
+
+        None (default)
+            Does not exclude any edge.
+        'reverse_id'
+            Exclude all edges specified in ``eids``, as well as their reverse edges
+            of the same edge type.
+
+            The mapping from each edge ID to its reverse edge ID is specified in
+            the keyword argument ``reverse_eid_map``.
+
+            This mode assumes that the reverse of an edge with ID ``e`` and type
+            ``etype`` will have ID ``reverse_eid_map[e]`` and type ``etype``.
+        'reverse_types'
+            Exclude all edges specified in ``eids``, as well as their reverse
+            edges of the corresponding edge types.
+
+            The mapping from each edge type to its reverse edge type is specified
+            in the keyword argument ``reverse_etype_map``.
+
+            This mode assumes that the reverse of an edge with ID ``e`` and type ``etype``
+            will have ID ``e`` and type ``reverse_etype_map[etype]``.
+    eids : Tensor or dict[etype, Tensor]
+        The edge IDs.
+    reverse_eid_map : Tensor or dict[etype, Tensor]
+        The mapping from edge ID to its reverse edge ID.
+    reverse_etype_map : dict[etype, etype]
+        The mapping from edge etype to its reverse edge type.
+    """
+    if exclude_mode is None:
+        return None
+    elif exclude_mode == 'reverse_id':
+        return _find_exclude_eids_with_reverse_id(g, eids, kwargs['reverse_eid_map'])
+    elif exclude_mode == 'reverse_types':
+        return _find_exclude_eids_with_reverse_types(g, eids, kwargs['reverse_etype_map'])
+    else:
+        raise ValueError('unsupported mode {}'.format(exclude_mode))
+
+
+class BlockSampler(object):
+    """Abstract class specifying the neighborhood sampling strategy for DGL data loaders.
+
+    The main method for BlockSampler is :func:`~dgl.dataloading.BlockSampler.sample_blocks`,
+    which generates a list of blocks for a multi-layer GNN given a set of seed nodes to
+    have their outputs computed.
+
+    The default implementation of :py:meth:`~dgl.dataloading.BlockSampler.sample_blocks` is
+    to repeat ``num_hops`` times the following:
+
+    * Obtain a frontier with the same nodes as the original graph but only the edges
+      involved in message passing on the last layer.
+      Customizable via :py:meth:`~dgl.dataloading.BlockSampler.sample_frontier`.
+
+    * Optionally, if the task is link prediction or edge classfication, remove edges
+      connecting training node pairs.  If the graph is undirected, also remove the
+      reverse edges.  This is controlled by the argument ``exclude_eids`` in
+      :py:meth:``~dgl.dataloading.BlockSampler.sample_blocks`` method.
+
+    * Convert the frontier into a block.
+
+    * Optionally assign the edge IDs to the block, controlled by the argument
+      ``return_eids`` in :py:meth:``~dgl.dataloading.BlockSampler.sample_blocks`` method.
```

Comment:
They are referring to the edge IDs in the original graph.  This is needed because users want to use edge features for minibatch training (requested in multiple threads in discussion forums).  The necessity is partially negated because of the recent subframe PR, but it still makes sense if users indeed need the IDs.  We don't want to have it on because it takes time to actually obtain the edge IDs in the original graph; we want to make sampling as light as possible.